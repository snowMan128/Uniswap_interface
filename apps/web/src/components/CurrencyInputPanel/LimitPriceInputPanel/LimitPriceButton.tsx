import { Trans } from '@lingui/macro'
import { Percent } from '@uniswap/sdk-core'
import Row from 'components/Row'
import { X } from 'react-feather'
import styled, { css } from 'styled-components'
import { ClickableStyle } from 'theme/components'
import { useFormatter } from 'utils/formatNumbers'

interface LimitPriceButtonProps {
  priceAdjustmentPercentage: number
  disabled?: boolean
  selected?: boolean
  onSelect: (priceAdjustmentPercentage: number) => void
}

const containerBorderCss = css`
  border-radius: 999px;
  border: 1px solid ${({ theme }) => theme.surface3};
`

const highlightedBorderCss = css`
  border-radius: 999px 0px 0px 999px;
  border-top: 1px solid ${({ theme }) => theme.surface3};
  border-bottom: 1px solid ${({ theme }) => theme.surface3};
  border-left: 1px solid ${({ theme }) => theme.surface3};
  border-right: 1px solid transparent;
`

const Container = styled.button<{ $selected?: boolean; $disabled?: boolean; $highlighted: boolean }>`
  background-color: ${({ theme, $highlighted }) => ($highlighted ? theme.surface3 : 'unset')};
  color: ${({ theme, $selected, $disabled, $highlighted }) =>
    $highlighted || ($selected && !$disabled) ? theme.neutral1 : theme.neutral2};
  padding: 2px 8px;
  height: 24px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme, $selected, $highlighted }) =>
    $selected || $highlighted ? theme.surface3 : 'transparent'};
  ${({ $highlighted }) => ($highlighted ? highlightedBorderCss : containerBorderCss)};
  ${({ $disabled }) => !$disabled && ClickableStyle};
`

const HighlightedContainerXButton = styled.button`
  ${ClickableStyle}
  background-color: ${({ theme }) => theme.surface3};
  color: ${({ theme }) => theme.neutral2};
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
  padding: 6px 6px 6px 4px;
  height: 24px;
  border-radius: 0px 999px 999px 0px;
  border-top: 1px solid ${({ theme }) => theme.surface3};
  border-bottom: 1px solid ${({ theme }) => theme.surface3};
  border-right: 1px solid ${({ theme }) => theme.surface3};
  border-left: 1px solid transparent;
`

export function LimitPresetPriceButton({
  priceAdjustmentPercentage,
  selected,
  disabled,
  onSelect,
}: LimitPriceButtonProps) {
  const { formatPercent } = useFormatter()
  return (
    <Container
      $selected={selected}
      $disabled={disabled}
      $highlighted={false}
      onClick={() => !disabled && onSelect(priceAdjustmentPercentage)}
    >
      {priceAdjustmentPercentage === 0 ? (
        <Trans>Current</Trans>
      ) : (
        <span>+{formatPercent(new Percent(priceAdjustmentPercentage, 100))}</span>
      )}
    </Container>
  )
}

/**
 * A button to reset the price to the market price (i.e. an adjustment of 0%)
 * When defined, this button displays customAdjustmentPercentage instead of "Current"
 */
export function LimitCustomMarketPriceButton({
  customAdjustmentPercentage,
  selected,
  disabled,
  onSelect,
}: Omit<LimitPriceButtonProps, 'priceAdjustmentPercentage'> & { customAdjustmentPercentage?: number }) {
  const onSetAdjustmentPercentage = () => !disabled && onSelect(0)
  const { formatPercent } = useFormatter()

  return (
    <Row width="unset" gap="1px">
      <Container
        $selected={selected}
        $disabled={disabled}
        $highlighted={customAdjustmentPercentage !== undefined}
        onClick={onSetAdjustmentPercentage}
      >
        {!customAdjustmentPercentage ? (
          <Trans>Current</Trans>
        ) : (
          <span>
            {customAdjustmentPercentage > 0 ? '+' : ''}
            {formatPercent(new Percent(customAdjustmentPercentage, 100))}
          </span>
        )}
      </Container>
      {customAdjustmentPercentage && (
        <HighlightedContainerXButton onClick={onSetAdjustmentPercentage}>
          <X size={12} />
        </HighlightedContainerXButton>
      )}
    </Row>
  )
}
