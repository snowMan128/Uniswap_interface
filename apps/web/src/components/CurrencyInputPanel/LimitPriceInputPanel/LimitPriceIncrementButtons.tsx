import Row from 'components/Row'
import { Minus, Plus } from 'react-feather'
import styled from 'styled-components'
import { ClickableStyle } from 'theme/components'

const Container = styled(Row)`
  width: fit-content;
  button {
    display: flex;
    align-items: center;
    justify-content: center;
    color: ${({ theme }) => theme.neutral2};
    background-color: ${({ theme }) => theme.surface3};
    border: 1px solid transparent;
    ${ClickableStyle}
    height: 24px;
    padding-bottom: 2px;
  }
`

const IncrementButton = styled.button`
  border-radius: 0px 999px 999px 0px;
  padding-right: 10px;
`

const DecrementButton = styled.button`
  padding-left: 10px;
  border-radius: 999px 0px 0px 999px;
`

export function LimitPriceIncrementButtons({
  onDecrement,
  onIncrement,
}: {
  onDecrement: () => void
  onIncrement: () => void
}) {
  return (
    <Container gap="1px">
      <DecrementButton onClick={onDecrement}>
        <Minus size={12} />
      </DecrementButton>
      <IncrementButton onClick={onIncrement}>
        <Plus size={12} />
      </IncrementButton>
    </Container>
  )
}
