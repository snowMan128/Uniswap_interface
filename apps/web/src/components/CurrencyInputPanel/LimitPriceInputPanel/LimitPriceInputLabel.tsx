import { Trans } from '@lingui/macro'
import { Currency } from '@uniswap/sdk-core'
import CurrencyLogo from 'components/Logo/CurrencyLogo'
import Row from 'components/Row'
import styled from 'styled-components'
import { ThemedText } from 'theme/components'

const InputCurencySymbolContainer = styled.span`
  display: inline-block;
  margin: 0 8px;
`

export function LimitPriceInputLabel({
  inputCurrency,
  showCurrencyMessage,
}: {
  inputCurrency?: Currency
  showCurrencyMessage: boolean
}) {
  if (!inputCurrency || !showCurrencyMessage) {
    return (
      <ThemedText.LabelSmall style={{ userSelect: 'none' }}>
        <Trans>Limit price</Trans>
      </ThemedText.LabelSmall>
    )
  }
  return (
    <ThemedText.LabelSmall style={{ userSelect: 'none' }}>
      <Row align="center">
        <Trans>When 1</Trans>{' '}
        <InputCurencySymbolContainer>
          <Row gap="xs" align="center" height="100%">
            <CurrencyLogo currency={inputCurrency} size="16px" />
            <ThemedText.BodyPrimary display="inline">{inputCurrency.symbol}</ThemedText.BodyPrimary>
          </Row>
        </InputCurencySymbolContainer>{' '}
        <Trans>is worth</Trans>
      </Row>
    </ThemedText.LabelSmall>
  )
}
