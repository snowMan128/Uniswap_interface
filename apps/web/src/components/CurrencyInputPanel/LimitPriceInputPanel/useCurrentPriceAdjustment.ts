import { Currency, CurrencyAmount, Fraction, Price } from '@uniswap/sdk-core'
import { parseUnits } from 'ethers/lib/utils'
import JSBI from 'jsbi'
import { useMemo } from 'react'

type CurrentPriceAdjustmentResult =
  | { currentPriceAdjustment: number; priceError: boolean }
  | { currentPriceAdjustment: undefined; priceError: false }

export function useCurrentPriceAdjustment({
  parsedLimitPrice,
  marketPrice,
  baseCurrency,
  quoteCurrency,
  limitPriceInverted,
}: {
  parsedLimitPrice?: Price<Currency, Currency>
  marketPrice?: Price<Currency, Currency>
  baseCurrency?: Currency
  quoteCurrency?: Currency
  limitPriceInverted: boolean
}): CurrentPriceAdjustmentResult {
  return useMemo(() => {
    if (
      !parsedLimitPrice ||
      !marketPrice ||
      !baseCurrency ||
      !quoteCurrency ||
      !parsedLimitPrice.baseCurrency.equals(baseCurrency)
    ) {
      return { currentPriceAdjustment: undefined, priceError: false }
    }
    const oneUnitOfBaseCurrency = CurrencyAmount.fromRawAmount(
      baseCurrency,
      JSBI.BigInt(parseUnits('1', baseCurrency?.decimals))
    )

    const marketQuote = marketPrice.quote(oneUnitOfBaseCurrency)
    const parsedPriceQuote = parsedLimitPrice.quote(oneUnitOfBaseCurrency)
    const decimalsScalar = JSBI.exponentiate(JSBI.BigInt(10), JSBI.BigInt(18))

    const scaledMarketQuote = JSBI.BigInt(marketQuote.multiply(decimalsScalar).toFixed(0))
    const scaledPriceQuote = JSBI.BigInt(parsedPriceQuote.multiply(decimalsScalar).toFixed(0))

    const difference = JSBI.subtract(scaledPriceQuote, scaledMarketQuote)
    const percentageChange = new Fraction(difference, scaledMarketQuote)

    const currentPriceAdjustment = Math.round(Number(percentageChange.multiply(100).toFixed(2)))
    return {
      currentPriceAdjustment,
      priceError: limitPriceInverted ? (currentPriceAdjustment ?? 0) > 0 : (currentPriceAdjustment ?? 0) < 0,
    }
  }, [parsedLimitPrice, marketPrice, baseCurrency, quoteCurrency, limitPriceInverted])
}
