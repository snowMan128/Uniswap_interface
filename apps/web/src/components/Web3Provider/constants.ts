import GNOSIS_ICON from 'assets/images/gnosis.png'
import COINBASE_ICON from 'assets/wallets/coinbase-icon.svg'
import METAMASK_ICON from 'assets/wallets/metamask-icon.svg'
import UNIWALLET_ICON from 'assets/wallets/uniswap-wallet-icon.png'
import WALLET_CONNECT_ICON from 'assets/wallets/walletconnect-icon.svg'
import { atomWithStorage, useAtomValue } from 'jotai/utils'

export const CONNECTION = {
  WALLET_CONNECT_CONNECTOR_ID: 'walletConnect',
  UNISWAP_WALLET_CONNECT_CONNECTOR_ID: 'uniswapWalletConnect',
  INJECTED_CONNECTOR_ID: 'injected',
  INJECTED_CONNECTOR_TYPE: 'injected',
  COINBASE_SDK_CONNECTOR_ID: 'coinbaseWalletSDK',
  COINBASE_RDNS: 'com.coinbase.wallet',
  METAMASK_RDNS: 'io.metamask',
  UNISWAP_EXTENSION_RDNS: 'org.uniswap.app',
  SAFE_CONNECTOR_ID: 'safe',
} as const

export const CONNECTOR_ICON_OVERRIDE_MAP: { [id in string]?: string } = {
  [CONNECTION.METAMASK_RDNS]: METAMASK_ICON,
  [CONNECTION.UNISWAP_WALLET_CONNECT_CONNECTOR_ID]: UNIWALLET_ICON,
  [CONNECTION.COINBASE_SDK_CONNECTOR_ID]: COINBASE_ICON,
  [CONNECTION.WALLET_CONNECT_CONNECTOR_ID]: WALLET_CONNECT_ICON,
  [CONNECTION.SAFE_CONNECTOR_ID]: GNOSIS_ICON,
}

// Used to track which connector was used most recently for UI states.
export const recentConnectorIdAtom = atomWithStorage<string | undefined>('recentConnectorId', undefined)
export function useRecentConnectorId() {
  return useAtomValue(recentConnectorIdAtom)
}
