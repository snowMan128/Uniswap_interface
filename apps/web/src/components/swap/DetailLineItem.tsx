import { LoadingRow } from 'components/Loader/styled'
import { RowBetween } from 'components/Row'
import { MouseoverTooltip, TooltipSize } from 'components/Tooltip'
import useHoverProps from 'hooks/useHoverProps'
import { useIsMobile } from 'nft/hooks'
import { PropsWithChildren } from 'react'
import styled from 'styled-components'
import { ThemedText } from 'theme/components'

export type LineItemData = {
  Label: React.FC
  Value: React.FC
  TooltipBody?: React.FC
  tooltipSize?: TooltipSize
  loaderWidth?: number
}

const LabelText = styled(ThemedText.BodySmall)<{ hasTooltip?: boolean }>`
  cursor: ${({ hasTooltip }) => (hasTooltip ? 'help' : 'auto')};
  color: ${({ theme }) => theme.neutral2};
`

const DetailRowValue = styled(ThemedText.BodySmall)`
  text-align: right;
  overflow-wrap: break-word;
`

type ValueWrapperProps = PropsWithChildren<{
  lineItem: LineItemData
  labelHovered: boolean
  syncing: boolean
}>

function ValueWrapper({ children, lineItem, labelHovered, syncing }: ValueWrapperProps) {
  const { TooltipBody, tooltipSize, loaderWidth } = lineItem
  const isMobile = useIsMobile()

  if (syncing) return <LoadingRow data-testid="loading-row" height={15} width={loaderWidth ?? 50} />

  if (!TooltipBody) return <DetailRowValue>{children}</DetailRowValue>

  return (
    <MouseoverTooltip
      placement={isMobile ? 'auto' : 'right'}
      forceShow={labelHovered} // displays tooltip when hovering either both label or value
      size={tooltipSize}
      text={
        <ThemedText.Caption color="neutral2">
          <TooltipBody />
        </ThemedText.Caption>
      }
    >
      <DetailRowValue>{children}</DetailRowValue>
    </MouseoverTooltip>
  )
}

export function DetailLineItem({ LineItem, syncing }: { LineItem: LineItemData; syncing?: boolean }) {
  const [labelHovered, hoverProps] = useHoverProps()
  return (
    <RowBetween>
      <LabelText {...hoverProps} hasTooltip={!!LineItem.TooltipBody} data-testid="swap-li-label">
        <LineItem.Label />
      </LabelText>
      <ValueWrapper lineItem={LineItem} labelHovered={labelHovered} syncing={syncing ?? false}>
        <LineItem.Value />
      </ValueWrapper>
    </RowBetween>
  )
}
