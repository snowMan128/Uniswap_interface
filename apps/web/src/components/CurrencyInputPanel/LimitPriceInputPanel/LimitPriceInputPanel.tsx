import { Currency, CurrencyAmount, Fraction, Price } from '@uniswap/sdk-core'
import { InputPanel } from 'components/CurrencyInputPanel/SwapCurrencyInputPanel'
import CurrencyLogo from 'components/Logo/CurrencyLogo'
import { StyledNumericalInput } from 'components/NumericalInput'
import Row from 'components/Row'
import { parseUnits } from 'ethers/lib/utils'
import JSBI from 'jsbi'
import { useCallback, useMemo } from 'react'
import { useLimitContext, useLimitPrice } from 'state/limit/LimitContext'
import { useSwapAndLimitContext } from 'state/swap/SwapContext'
import styled from 'styled-components'
import { ThemedText } from 'theme/components'
import { NumberType, useFormatter } from 'utils/formatNumbers'

import { formatCurrencySymbol } from '../utils'
import { LimitCustomMarketPriceButton, LimitPresetPriceButton } from './LimitPriceButton'
import { LimitPriceInputLabel } from './LimitPriceInputLabel'

const OutputCurrencyContainer = styled(Row)`
  user-select: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`

const TextInputRow = styled.div`
  display: flex;
  flex-grow: 1;
`

const PRICE_ADJUSTMENT_PRESETS = [1, 5, 10]

export function LimitPriceInputPanel() {
  const { limitPrice, setLimitPrice } = useLimitPrice()
  const {
    derivedLimitInfo: { parsedLimitPrice, marketPrice },
  } = useLimitContext()

  const {
    currencyState: { inputCurrency, outputCurrency },
  } = useSwapAndLimitContext()

  const { formatCurrencyAmount } = useFormatter()

  const formattedLimitPriceOutputAmount: string = useMemo(() => {
    // If the user has manually typed in a limit price, use that.
    if (limitPrice) return limitPrice
    // Otherwise, use parsedLimitPrice which may have been calculated from input/output amounts.
    if (!inputCurrency) return ''
    return formatCurrencyAmount({
      amount: parsedLimitPrice?.quote(CurrencyAmount.fromRawAmount(inputCurrency, 1)),
      type: NumberType.SwapTradeAmount,
      placeholder: '',
    })
  }, [limitPrice, inputCurrency, formatCurrencyAmount, parsedLimitPrice])

  const adjustedPrices = useMemo(() => {
    if (!marketPrice || !inputCurrency || !outputCurrency) return undefined
    const getAdjustedPrice = (priceAdjustmentPercentage: number) => {
      return new Price({
        // 100 input token
        baseAmount: CurrencyAmount.fromRawAmount(inputCurrency, JSBI.BigInt(parseUnits('100', inputCurrency.decimals))),
        // (100 + adjustmentPercentage) times the market quote amount for 1 input token
        quoteAmount: CurrencyAmount.fromRawAmount(
          outputCurrency,
          JSBI.multiply(
            JSBI.BigInt(100 + priceAdjustmentPercentage),
            marketPrice.quote(
              CurrencyAmount.fromRawAmount(inputCurrency, JSBI.BigInt(parseUnits('1', inputCurrency.decimals)))
            ).quotient
          )
        ),
      })
    }
    return {
      1: getAdjustedPrice(1),
      5: getAdjustedPrice(5),
      10: getAdjustedPrice(10),
    }
  }, [inputCurrency, marketPrice, outputCurrency])

  const currentPriceAdjustment: number | undefined = useMemo(() => {
    if (!parsedLimitPrice || !marketPrice || !inputCurrency || !outputCurrency) return undefined
    const marketQuote = marketPrice.quote(
      CurrencyAmount.fromRawAmount(inputCurrency, 10 ** inputCurrency.decimals)
    ).quotient
    const parsedPriceQuote = parsedLimitPrice.quote(
      CurrencyAmount.fromRawAmount(inputCurrency, 10 ** inputCurrency.decimals)
    ).quotient
    const difference = JSBI.subtract(parsedPriceQuote, marketQuote)
    const percentageChange = new Fraction(difference, marketQuote)

    return Number(percentageChange.toFixed(2)) * 100
  }, [inputCurrency, outputCurrency, marketPrice, parsedLimitPrice])

  const onSelectLimitPrice = useCallback(
    (adjustedPrice: Price<Currency, Currency> | undefined) => {
      if (!inputCurrency) return
      const marketOutputAmount = adjustedPrice?.quote(
        CurrencyAmount.fromRawAmount(inputCurrency, 10 ** inputCurrency.decimals)
      )
      setLimitPrice(
        formatCurrencyAmount({
          amount: marketOutputAmount,
          type: NumberType.SwapTradeAmount,
          placeholder: limitPrice,
        })
      )
    },
    [formatCurrencyAmount, inputCurrency, limitPrice, setLimitPrice]
  )

  return (
    <InputPanel>
      <LimitPriceInputLabel inputCurrency={inputCurrency} showCurrencyMessage={!!formattedLimitPriceOutputAmount} />
      <TextInputRow>
        <StyledNumericalInput
          disabled={!(inputCurrency && outputCurrency)}
          className="limit-price-input"
          value={formattedLimitPriceOutputAmount}
          onUserInput={setLimitPrice}
          $loading={false}
        />
        {outputCurrency && (
          <OutputCurrencyContainer gap="xs" width="unset">
            <CurrencyLogo currency={outputCurrency} size="16px" />
            <ThemedText.BodyPrimary className="token-symbol-container">
              {formatCurrencySymbol(outputCurrency)}
            </ThemedText.BodyPrimary>
          </OutputCurrencyContainer>
        )}
      </TextInputRow>
      <Row gap="sm" marginTop="8px">
        <LimitCustomMarketPriceButton
          key="limit-price-market"
          customAdjustmentPercentage={
            currentPriceAdjustment !== undefined &&
            currentPriceAdjustment !== 0 &&
            !PRICE_ADJUSTMENT_PRESETS.includes(currentPriceAdjustment)
              ? currentPriceAdjustment
              : undefined
          }
          disabled={!inputCurrency || !outputCurrency}
          selected={Boolean(
            currentPriceAdjustment !== undefined && !PRICE_ADJUSTMENT_PRESETS.includes(currentPriceAdjustment)
          )}
          onSelect={() => onSelectLimitPrice(marketPrice)}
        />
        {PRICE_ADJUSTMENT_PRESETS.map((adjustmentPercentage) => {
          const adjustedPrice = adjustedPrices?.[adjustmentPercentage as keyof typeof adjustedPrices]
          return (
            <LimitPresetPriceButton
              key={`limit-price-${adjustmentPercentage}`}
              priceAdjustmentPercentage={adjustmentPercentage}
              disabled={!inputCurrency || !outputCurrency || !marketPrice}
              // TODO (WEB-3416): give this equality check some small +- threshold
              selected={currentPriceAdjustment === adjustmentPercentage}
              onSelect={() => onSelectLimitPrice(adjustedPrice)}
            />
          )
        })}
      </Row>
    </InputPanel>
  )
}
