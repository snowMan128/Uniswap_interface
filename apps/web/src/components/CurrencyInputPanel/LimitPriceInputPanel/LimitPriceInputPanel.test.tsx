import { SwapTab } from 'components/swap/constants'
import { DAI, USDC_MAINNET } from 'constants/tokens'
import { LimitContext } from 'state/limit/LimitContext'
import { SwapAndLimitContext } from 'state/swap/SwapContext'
import { render, screen } from 'test-utils/render'

import { LimitPriceInputPanel } from './LimitPriceInputPanel'

const mockSwapAndLimitContextValue = {
  currencyState: {
    inputCurrency: DAI,
    outputCurrency: undefined,
  },
  prefilledState: {},
  setCurrencyState: jest.fn(),
  currentTab: SwapTab.Limit,
  setCurrentTab: jest.fn(),
}

const mockLimitContextValue = {
  limitState: {
    inputAmount: '',
    limitPrice: '100',
    outputAmount: '',
    expiry: 1,
    isInputAmountFixed: true,
  },
  setLimitState: jest.fn(),
  derivedLimitInfo: {
    currencyBalances: {},
    parsedAmounts: {},
  },
}

describe('LimitPriceInputPanel', () => {
  it('should render the component with no currencies selected', () => {
    const { container } = render(<LimitPriceInputPanel />)
    expect(screen.getByText('Limit price')).toBeVisible()
    expect(screen.getByPlaceholderText('0')).toBeVisible()
    expect(screen.getByText('Current')).toBeVisible()
    expect(screen.getByText('+1%')).toBeVisible()
    expect(screen.getByText('+5%')).toBeVisible()
    expect(screen.getByText('+10%')).toBeVisible()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render correct subheader with inputCurrency defined, but no price', () => {
    const { container } = render(
      <SwapAndLimitContext.Provider value={mockSwapAndLimitContextValue}>
        <LimitPriceInputPanel />
      </SwapAndLimitContext.Provider>
    )
    expect(screen.getByText('Limit price')).toBeVisible()
    expect(screen.getByPlaceholderText('0')).toBeVisible()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render correct subheader with input currency and limit price defined', () => {
    render(
      <SwapAndLimitContext.Provider value={mockSwapAndLimitContextValue}>
        <LimitContext.Provider value={mockLimitContextValue}>
          <LimitPriceInputPanel />
        </LimitContext.Provider>
      </SwapAndLimitContext.Provider>
    )
    expect(screen.getByText('DAI')).toBeVisible()
    expect(screen.getByPlaceholderText('0')).toBeVisible()
  })

  it('should render the output currency when defined', () => {
    const { container } = render(
      <SwapAndLimitContext.Provider
        value={{
          ...mockSwapAndLimitContextValue,
          currencyState: {
            ...mockSwapAndLimitContextValue.currencyState,
            outputCurrency: USDC_MAINNET,
          },
        }}
      >
        <LimitContext.Provider value={mockLimitContextValue}>
          <LimitPriceInputPanel />
        </LimitContext.Provider>
      </SwapAndLimitContext.Provider>
    )
    expect(screen.getByText('DAI')).toBeVisible() // subheader
    expect(container.querySelector('.token-symbol-container')).toHaveTextContent('USDC')
    expect(screen.getByPlaceholderText('0')).toBeVisible()
  })
})
