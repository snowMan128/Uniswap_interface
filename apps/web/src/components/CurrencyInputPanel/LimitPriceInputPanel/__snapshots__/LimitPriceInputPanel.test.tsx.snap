// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitPriceInputPanel should render correct subheader with inputCurrency defined, but no price 1`] = `
.c5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 8px;
}

.c7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: unset;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 8px;
}

.c8 {
  width: unset;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 1px;
}

.c1 {
  color: #7D7D7D;
  -webkit-letter-spacing: -0.01em;
  -moz-letter-spacing: -0.01em;
  -ms-letter-spacing: -0.01em;
  letter-spacing: -0.01em;
}

.c3 {
  color: #222222;
  pointer-events: none;
  width: 0;
  position: relative;
  font-weight: 485;
  outline: none;
  border: none;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  background-color: transparent;
  font-size: 28px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0px;
  -webkit-appearance: textfield;
  text-align: right;
}

.c3::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c3 [type='number'] {
  -moz-appearance: textfield;
}

.c3::-webkit-outer-spin-button,
.c3::-webkit-inner-spin-button {
  -webkit-appearance: none;
}

.c3::-webkit-input-placeholder {
  color: #CECECE;
}

.c3::-moz-placeholder {
  color: #CECECE;
}

.c3:-ms-input-placeholder {
  color: #CECECE;
}

.c3::placeholder {
  color: #CECECE;
}

.c4 {
  -webkit-filter: none;
  filter: none;
  opacity: 1;
  -webkit-transition: opacity 250ms ease-in-out;
  transition: opacity 250ms ease-in-out;
  text-align: left;
  font-size: 36px;
  font-weight: 485;
  max-height: 44px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  border-radius: 20px;
  z-index: 1;
  width: initial;
  -webkit-transition: height 1s ease;
  transition: height 1s ease;
  will-change: height;
}

.c9 {
  background-color: unset;
  color: #7D7D7D;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border-radius: 999px;
  border: 1px solid #22222212;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  class="c0"
>
  <div
    class="c1 css-142zc9n"
    style="user-select: none;"
  >
    Limit price
  </div>
  <div
    class="c2"
  >
    <input
      autocomplete="off"
      autocorrect="off"
      class="c3 c4 limit-price-input"
      disabled=""
      inputmode="decimal"
      maxlength="79"
      minlength="1"
      pattern="^[0-9]*[.,]?[0-9]*$"
      placeholder="0"
      spellcheck="false"
      type="text"
      value=""
    />
  </div>
  <div
    class="c5 c6"
  >
    <div
      class="c7 c8"
      width="unset"
    >
      <button
        class="c9"
      >
        Current
      </button>
    </div>
    <button
      class="c9"
    >
      <span>
        +
        1%
      </span>
    </button>
    <button
      class="c9"
    >
      <span>
        +
        5%
      </span>
    </button>
    <button
      class="c9"
    >
      <span>
        +
        10%
      </span>
    </button>
  </div>
</div>
`;

exports[`LimitPriceInputPanel should render the component with no currencies selected 1`] = `
.c5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 8px;
}

.c7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: unset;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 8px;
}

.c8 {
  width: unset;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 1px;
}

.c1 {
  color: #7D7D7D;
  -webkit-letter-spacing: -0.01em;
  -moz-letter-spacing: -0.01em;
  -ms-letter-spacing: -0.01em;
  letter-spacing: -0.01em;
}

.c3 {
  color: #222222;
  pointer-events: none;
  width: 0;
  position: relative;
  font-weight: 485;
  outline: none;
  border: none;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  background-color: transparent;
  font-size: 28px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0px;
  -webkit-appearance: textfield;
  text-align: right;
}

.c3::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c3 [type='number'] {
  -moz-appearance: textfield;
}

.c3::-webkit-outer-spin-button,
.c3::-webkit-inner-spin-button {
  -webkit-appearance: none;
}

.c3::-webkit-input-placeholder {
  color: #CECECE;
}

.c3::-moz-placeholder {
  color: #CECECE;
}

.c3:-ms-input-placeholder {
  color: #CECECE;
}

.c3::placeholder {
  color: #CECECE;
}

.c4 {
  -webkit-filter: none;
  filter: none;
  opacity: 1;
  -webkit-transition: opacity 250ms ease-in-out;
  transition: opacity 250ms ease-in-out;
  text-align: left;
  font-size: 36px;
  font-weight: 485;
  max-height: 44px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  border-radius: 20px;
  z-index: 1;
  width: initial;
  -webkit-transition: height 1s ease;
  transition: height 1s ease;
  will-change: height;
}

.c9 {
  background-color: unset;
  color: #7D7D7D;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border-radius: 999px;
  border: 1px solid #22222212;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  class="c0"
>
  <div
    class="c1 css-142zc9n"
    style="user-select: none;"
  >
    Limit price
  </div>
  <div
    class="c2"
  >
    <input
      autocomplete="off"
      autocorrect="off"
      class="c3 c4 limit-price-input"
      disabled=""
      inputmode="decimal"
      maxlength="79"
      minlength="1"
      pattern="^[0-9]*[.,]?[0-9]*$"
      placeholder="0"
      spellcheck="false"
      type="text"
      value=""
    />
  </div>
  <div
    class="c5 c6"
  >
    <div
      class="c7 c8"
      width="unset"
    >
      <button
        class="c9"
      >
        Current
      </button>
    </div>
    <button
      class="c9"
    >
      <span>
        +
        1%
      </span>
    </button>
    <button
      class="c9"
    >
      <span>
        +
        5%
      </span>
    </button>
    <button
      class="c9"
    >
      <span>
        +
        10%
      </span>
    </button>
  </div>
</div>
`;
